### PR TITLE
Add more Conway tests

### PR DESCRIPTION
**Pull request by sgillespie**
# Description

Addresses a subset of https://github.com/input-output-hk/cardano-db-sync/issues/1526. In particular, I've added the following test groups:

 * rollbacks
 * different configs
 * blocks with txs
 * stake addresses
 * stake distribution
 * rewards

# Checklist

- [x] Commit sequence broadly makes sense
- [x] Commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [ ] Any changes are noted in the [changelog](https://github.com/input-output-hk/cardano-db-sync/blob/master/cardano-db-sync/CHANGELOG.md)
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) on version 0.10.1.0 (which can be run with `scripts/fourmolize.sh`)
- [x] Self-reviewed the diff

# Migrations

- [ ] The pr causes a [breaking change](https://github.com/input-output-hk/cardano-db-sync/blob/master/doc/migrations.md) of type a,b or c
- [ ] If there is a breaking change, the pr includes a database migration and/or a fix process for old values, so that upgrade is possible
- [ ] Resyncing and running the migrations provided will result in the same database semantically

If there is a breaking change, especially a big one, please add a justification here. Please elaborate
more what the migration achieves, what it cannot achieve or why a migration is not possible.

### CODE REVIEW

File: cardano-chain-gen/test/Test/Cardano/Db/Mock/Unit/Conway/CommandLineArg/ForceIndex.hs [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+import Test.Tasty.HUnit (Assertion ())
+import Prelude ()
+
+checkForceIndexesArg :: IOManager -> [(Text, Text)] -> Assertion
+checkForceIndexesArg =
+  withCustomConfig cliArgs conwayConfigDir testLabel $ \_ _ dbSync -> do
+    startDBSync dbSync
+
+    -- Verify number of DB indexes
+    assertEqQuery dbSync DB.queryPgIndexesCount 159 "unexpected number of indexes"
```
Comment (kderme) [inline cardano-chain-gen/test/Test/Cardano/Db/Mock/Unit/Conway/CommandLineArg/ForceIndex.hs:20]: Fyi after a rebase this value may need to change to 161

File: cardano-chain-gen/test/Test/Cardano/Db/Mock/Unit/Conway/CommandLineArg/ForceIndex.hs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    cliArgs = initCommandLineArgs {claForceIndexes = True}
+    testLabel = "conwayCLACheckForceIndexesArg"
+
+checkNoForceIndexesArg :: IOManager -> [(Text, Text)] -> Assertion
+checkNoForceIndexesArg =
+  withCustomConfigAndDropDB cliArgs conwayConfigDir testLabel $ \_ _ dbSync -> do
+    startDBSync dbSync
+
+    -- Verify number of DB indexes
+    assertEqQuery dbSync DB.queryPgIndexesCount 94 "unexpected number of indexes"
```
Comment (kderme) [inline cardano-chain-gen/test/Test/Cardano/Db/Mock/Unit/Conway/CommandLineArg/ForceIndex.hs:31]: Same here. To 96

File: cardano-chain-gen/src/Cardano/Mock/Forging/Tx/Conway.hs [hunk 3]
```diff
@@ -124,6 +184,81 @@ mkSimpleTx isValid' txBody =
...
+  Delegatee StandardCrypto ->
+  StakeCredential StandardCrypto ->
+  ConwayTxCert StandardConway
+mkDelegTxCert delegatee = mkTxDelegCert (`ConwayDelegCert` delegatee)
+
+mkTxDelegCert ::
+  (StakeCredential StandardCrypto -> ConwayDelegCert StandardCrypto) ->
+  StakeCredential StandardCrypto ->
+  ConwayTxCert StandardConway
+mkTxDelegCert = (ConwayTxCertDeleg .)
```
Comment (Cmdv) [inline cardano-chain-gen/src/Cardano/Mock/Forging/Tx/Conway.hs:260]: minor change if you agree, just wondering if this might be a little too point free?

**Review by Cmdv — APPROVED** [review]
Really good, appreciate the comments in the tests cases too. Makes it much clearer what is happening 🎉 👌 :shipit:

File: cardano-chain-gen/src/Cardano/Mock/Forging/Tx/Conway.hs (hunk 3 above)
Comment (sgillespie) [reply cardano-chain-gen/src/Cardano/Mock/Forging/Tx/Conway.hs:260]: I went back and forth a few times, I don't mind making it a little more readable

Comment (Cmdv) [reply cardano-chain-gen/src/Cardano/Mock/Forging/Tx/Conway.hs:260]: no big deal we can leave it 👍

**Review by Cmdv — APPROVED** [review]
🎉

**Review by kderme — APPROVED** [review]
LGTM! Is there a list of tests that were not tranfered, like `rewardsDeregistration` or had to be altered for Conway?

**Comment by Cmdv** [comment]
@kderme I think he's still working through them but is going to break up the PR as it's getting rather large 😄

**Comment by sgillespie** [comment]
> LGTM! Is there a list of tests that were not tranfered, like rewardsDeregistration or had to be altered for Conway?

As far is I can remember, only tests from the rewards suite were not transferred:

 * rewardsDeregistration
 * rewardsReregistration
 * mirReward
 * mirRewardDereg
 * mirReward

Furthermore, any test that used deleg certificates had to be slightly altered, since Conway changed this. You'll see this in the rollback and stake suites, but there should be no change in overall behavior.

> I think he's still working through them but is going to break up the PR as it's getting rather large

This is also correct.  I still have the following test suites left:

 * plutus send scripts
 * plutus cert scripts
 * MultiAssets plutus scripts
 * pools and smash
 * Babbage inline and reference
 * Hard Fork

**Pull request merged**